contents/scalability/scalable-development.md

```diff
@@ -0,0 +1,63 @@
+# Scalable Development - An Introduction
+
+Authors: Loh Jia Shun Kenneth
+
+# Scalable Development
+When Pokémon Go first launched in July 6, the heavy traffic from users caused its servers to crash. However, as of 2017, Google handles at least 3.5 billion searches daily, Whatsapp handles at least 10 billion messages a day, and Facebook hosts 1.2 billion daily active users. What makes these companies different?
+
+The key difference is their server infrastructure. A scalable server allows a company to provide reliable service even as the user base increases. Even if your service might not have to cater to millions of users at the moment, designing scalable software now will reduce the need to restructure the entire backend when the need arises.
+
+# Scalability
+
+So what is scalability? Scalability is the capacity to which a system can cope with an increased workload. Workload can refer to number of users on the server, number of files uploaded, number of concurrent requests, or any quantity that the system needs to scale for.
+
+# Ways to scale
+
+Let's say you have a server running on a single host, but due to the growing number of users, it has been slowing down. What are the different ways you can speed it up?
+
+## Get A Better Host
+
+This is arguably the easiest way to speed up your server, as it is a general solution to most problems. Simply get faster CPUs, larger RAM, etc. This is called vertical scaling. If you get multi-core processors, use threading in your server to allow parallelization of certain processes to take advantage of the cores.
+
+However, due to Amdahl's law, each additional processor will give a less-than-proportionate speed up. Coupled with the exponential increase in price as the computing power of the CPU increases, this is typically not scalable in the long run.
```

Comment:
Amdahl's law applies in general cases.

For instance on Facebook a person's single update will result in notifications for all his/her friends. The serial portion is posting the event, saving to database, etc. Notifying friends will be the parallel portion.